File: object-oriented-programming/gcse-concepts/python/pet_game.py
# Raspberry Pi Foundation
# Developed as part of Ada Computer Science
# 
# Usage licensed under CC BY-NC-SA 4.0

import random

class Pet:

    # Constructor method
    def __init__(self, given_name, given_type, given_colour):
        self.name = given_name
        self.pet_type = given_type
        self.colour = given_colour
        self.sleeping = False

        # Randomly choose a starting mood
        all_moods = ["playful", "hungry", "tired"]
        self.mood = random.choice(all_moods)

    def get_name(self):
        return self.name

    def set_name(self, new_name):
        self.name = new_name

    def get_pet_type(self):
        return self.pet_type

    def set_pet_type(self, new_type):
        self.pet_type = new_type

    def get_colour(self):
        return self.colour

    def set_colour(self, new_colour):
        self.colour = new_colour

    def get_mood(self):
        return self.mood

    def is_sleeping(self):
        return self.sleeping

    def describe(self):
        print(f"I am a {self.mood}, {self.colour} {self.pet_type} called {self.name}")

    def play(self):
        """A method for trying to play with the pet"""
        if self.sleeping == True:
            print("Zzzzzzz. I am sleeping")
        elif self.mood == "hungry":
            print("I am too hungry to play")
        elif self.mood == "tired":
            print("I am too tired to play")
        else:
            print("This is fun, I love playing")

            # Randomly choose a new mood
            all_moods = ["playful", "hungry", "tired"]
            self.mood = random.choice(all_moods)

    def feed(self):
        """A method for trying to feed the pet"""
        if self.sleeping == True:
            print("Zzzzzzz. I am sleeping")
        elif self.mood == "tired":
            print("I am too sleepy to eat anything now")
        elif self.mood == "playful":
            print("I am not hungry - I want to play")
        else:
            print("Yum - yum - that tastes great")
            self.mood = "playful"

    def sleep(self):
        """A method for trying to put the pet to bed"""
        if self.mood == "playful":
            print("I am too playful to sleep")
        elif self.mood == "hungry":
            print("I need something to eat before I can go to sleep")
        else:
            print("Zzzzzzzz")
            self.sleeping = True

    def wake(self):
        """A method for trying to wake the pet up"""
        if self.sleeping == False:
           print("I wasn't even asleep!")
        else:
            print("I am awake and ready to play!")
            self.sleeping = False
            self.mood = "playful"



def change_pet(my_pet):
    """Allows pet to be changed"""

    # Ask the user to change the pet's name
    answer = input("\nDo you want to change the name of your pet(y/n)? ")
    if answer.lower() == "y":         
        new_name = input("Enter a new name for your pet ")
        my_pet.set_name(new_name)

    # Ask the user to change the pet's type
    answer = input("\nDo you want to change the type of pet you have(y/n)? ")
    if answer.lower() == "y":
        pet_name = my_pet.get_name()
        new_type = input(f"What type of animal is {pet_name}? ")
        my_pet.set_type(new_type)

    # Ask the user to change the pet's colour
    answer = input("\nDo you want to change the colour of your pet(y/n)? ")
    if answer.lower() == "y":
        pet_name = my_pet.get_name()
        new_colour = input(f"What colour is {pet_name}? ")
        my_pet.set_colour(new_colour)

    print()
    my_pet.describe()

def show_game_options():
    """Display options for playing game"""
    
    print()
    print(" 1. Feed your pet")
    print(" 2. Play with your pet")
    print(" 3. Put your pet to bed")
    print(" 4. Wake your pet up")
    print(" 5. Change pet")
    print(" 6. Finish the game")
    
def main():
    # Ask the user to input the pet's information
    pet_name = input("Enter a name for your pet: ")
    pet_type = input(f"What type of animal is {pet_name}? ")
    pet_colour = input(f"What colour is {pet_name}? ")

    # Make a new pet object
    my_pet = Pet(pet_name, pet_type, pet_colour)

    # Game loop for playing the game
    finished = False
    while finished == False:
        show_game_options()

        # Run a method depending on the user choice
        choice = input("What do you want to do? ")
        print()
        
        if choice == "1":
            my_pet.feed()
        elif choice == "2":
            my_pet.play()
        elif choice == "3":
            my_pet.sleep()
        elif choice == "4":
            my_pet.wake()
        elif choice == "5":
            change_pet(my_pet)
        elif choice == "6":
            # This will cause game to end
            finished = True
            print ("Thanks for playing the pet game")
        else:
            print("That is not a menu option")    

if __name__ == "__main__":
    main()

File: object-oriented-programming/overview/java/pet_game.java
/*
Raspberry Pi Foundation
Developed as part of Ada Computer Science
Usage licensed under CC BY-NC-SA 4.0

Note: This file is designed to be copied out and compiled on your machine.
In order for the program to compile properly in some IDEs, you need to ensure that the
class filename is the same as the class name with the main method in it.

To run this file you need to:
1. Copy the contents
2. Create a Java project in the IDE of your choice (Eclipse, for example)
3. Create a new Java Class that uses the same name as the class containing the main method in this program
4. Paste the contents into the new Java Class
5. Save and run the program
*/

import java.util.Random;
import java.util.Scanner;

class Pet
{
    private String name;
    private String petType;
    private String colour;
    private String mood;
    private boolean sleeping;

    // Constructor method
    public Pet(String givenName, String givenType, String givenColour)
    {
        name = givenName;
        petType = givenType;
        colour = givenColour;
        sleeping = false;

        // Randomly choose a starting mood
        String[] allMoods = { "playful", "hungry", "tired" };
        Random rnd = new Random();
        int index = rnd.nextInt(allMoods.length);
        mood = allMoods[index];
    }

    public String getName()
    {
        return name;
    }

    public void setName(String newName)
    {
        name = newName;
    }

    public String getPetType()
    {
        return petType;
    }

    public void setPetType(String newType)
    {
        petType = newType;
    }

    public String getColour()
    {
        return colour;
    }

    public void setColour(String newColour)
    {
        colour = newColour;
    }

    public String getMood()
    {
        return mood;
    }

    public boolean isSleeping()
    {
        return sleeping;
    }

    public void describe()
    {
        System.out.println("I am a " + mood + ", " + colour + " " + petType + " called " + name);
    }


    // A method for trying to play with the pet
    public void play()
    {
        if (sleeping == true) {
            System.out.println("Zzzzzzz. I am sleeping");
        }
        else if ("hungry".equals(mood)) {
            System.out.println("I am too hungry to play");
        }
        else if ("tired".equals(mood)) {
            System.out.println("I am too tired to play");
        }
        else {
            System.out.println("This is fun, I love playing");

            // Randomly choose a new mood
            String[] allMoods = { "playful", "hungry", "tired" };
            Random rnd = new Random();
            int index = rnd.nextInt(allMoods.length);
            mood = allMoods[index];
        }
    }


    // A method for trying to feed the pet
    public void feed()
    {
        if (sleeping == true) {
            System.out.println("Zzzzzzz. I am sleeping");
        }
        else if ("tired".equals(mood)) {
            System.out.println("I am too sleepy to eat anything now");
        }
        else if ("playful".equals(mood)) {
            System.out.println("I am not hungry - I want to play");
        }
        else {
            System.out.println("Yum - yum - that tastes great");
            mood = "playful";
        }
    }


    // A method for trying to put the pet to bed
    public void sleep()
    {
        if ("playful".equals(mood)) {
            System.out.println("I am too playful to sleep");
        }
        else if ("hungry".equals(mood)) {
            System.out.println("I need something to eat before I can go to sleep");
        }
        else {
            System.out.println("Zzzzzzzz");
            sleeping = true;
        }
    }


    // A method for trying to wake the pet up
    public void wake()
    {
        if (sleeping == false) {
            System.out.println("I wasn't even asleep!");
        }
        else {
            System.out.println("I am awake and ready to play!");
            sleeping = false;
            mood = "playful";
        }
    }


}


class TestingPetGame
{
    // The main method is the entry point for all Java programs
    public static void main(String[] args)
    {
        // Use Scanner for reading user input from the console
        Scanner scanner = new Scanner(System.in);
        
        // Ask the user to input the pet's information
        System.out.println("Enter a name for your pet:");
        String petName = scanner.nextLine();

        System.out.println("What type of animal is " + petName + "?");
        String petType = scanner.nextLine();

        System.out.println("What colour is " + petName + "?");
        String petColour = scanner.nextLine();

        // Make a new pet object
        Pet myPet = new Pet(petName, petType, petColour);

        // Game loop for playing the game
        boolean finished = false;
        while (finished == false) {
            showGameOptions();

            // Run a method depending on the user choice
            System.out.println("What do you want to do? ");
            String choice = scanner.nextLine();
            System.out.println();

            if ("1".equals(choice)) {
                myPet.feed();
            }
            else if ("2".equals(choice)) {
                myPet.play();
            }
            else if ("3".equals(choice)) {
                myPet.sleep();
            }
            else if ("4".equals(choice)) {
                myPet.wake();
            }
            else if ("5".equals(choice)) {
                changePet(myPet);
            }
            else if ("6".equals(choice)) {
                // This will cause game to end
                finished = true;
                System.out.println("Thanks for playing the pet game");
            }
            else {
                System.out.println("That is not a menu option");
            }
        }
    }


    // Display options for playing game
    public static void showGameOptions()
    {
        System.out.println();
        System.out.println(" 1. Feed your pet");
        System.out.println(" 2. Play with your pet");
        System.out.println(" 3. Put your pet to bed");
        System.out.println(" 4. Wake your pet up");
        System.out.println(" 5. Change pet");
        System.out.println(" 6. Finish the game");
    }


    // Allows pet to be changed
    public static void changePet(Pet myPet)
    {
        // Use Scanner for reading user input from the console
        Scanner scanner = new Scanner(System.in);

        // Ask the user to change the pet's name
        System.out.println("\nDo you want to change the name of your pet (y/n)?");
        String answer = scanner.nextLine();

        if ("y".equalsIgnoreCase(answer)) {
            System.out.println("Enter a new name for your pet ");
            String newName = scanner.nextLine();
            myPet.setName(newName);
        }

        // Ask the user to change the pet's type
        System.out.println("\nDo you want to change the type of pet you have (y/n)?");
        answer = scanner.nextLine();

        if ("y".equalsIgnoreCase(answer)) {
            String petName = myPet.getName();
            System.out.println("What type of animal is " + petName + "?");
            String newType = scanner.nextLine();
            myPet.setPetType(newType);
        }

        // Ask the user to change the pet's colour
        System.out.println("\nDo you want to change the colour of your pet (y/n)?");
        answer = scanner.nextLine();

        if ("y".equalsIgnoreCase(answer)) {
            String petName = myPet.getName();
            System.out.println("What colour is " + petName + "?");
            String newColour = scanner.nextLine();
            myPet.setColour(newColour);
        }

        System.out.println();
        myPet.describe();
    }


}